cmd/aks-periscope/aks-periscope.go

```diff
@@ -54,31 +104,103 @@ func main() {
+			networkOutboundCollector.GetName():         networkOutboundCollector,
+			ipTablesCollector.GetName():                ipTablesCollector,
+			nodeLogsCollector.GetName():                nodeLogsCollector,
+			dnsCollector.GetName():                     dnsCollector,
+			kubeObjectsCollector.GetName():             kubeObjectsCollector,
+			kubeletCmdCollector.GetName():              kubeletCmdCollector,
+			systemPerfCollector.GetName():              systemPerfCollector,
+			helmCollector.GetName():                    helmCollector,
+			osmCollector.GetName():                     osmCollector,
+		})
+
+	//diagnosers
+	//NOTE currently the collector instance objects are shared between the collector itself and things which use it as a dependency
+	networkConfigDiagnoser := diagnoser.NewNetworkConfigDiagnoser(dnsCollector, kubeletCmdCollector)
+	networkOutboundDiagnoser := diagnoser.NewNetworkOutboundDiagnoser(networkOutboundCollector)
+	selectedDiagnosers := selectDiagnosers(
+		map[string]interfaces.Diagnoser{
+			networkConfigDiagnoser.GetName():   networkConfigDiagnoser,
+			networkOutboundDiagnoser.GetName(): networkOutboundDiagnoser,
+		})
+
+	//exporters
+	azureBlobExporter := exporter.NewAzureBlobExporter(creationTimeStamp, hostname)
+	selectedExporters := selectExporters(
+		map[string]interfaces.Exporter{
+			azureBlobExporter.GetName():    azureBlobExporter,
+		})
+
+	return selectedCollectors, selectedDiagnosers, selectedExporters
+}
+
+// selectCollectors select the collectors to run by looking up the value from the config
+func selectCollectors(allCollectorsByName map[string]interfaces.Collector) []interfaces.Collector {
+	collectors := []interfaces.Collector{}
+
+	enabledCollectorString, found := os.LookupEnv("ENABLED_COLLECTORS")
+	if !found {
+		//if not defined, default to all collectors enabled
+		enabledCollectorString = "containerlogs dns helm iptables kubeletcmd kubeobjects networkoutbound nodelogs osm systemlogs systemperf"
 	}
 
-	if contains(collectorList, "connectedCluster") {
-		collectors = append(collectors, helmCollector)
-	} else {
-		collectors = append(collectors, systemLogsCollector)
-		collectors = append(collectors, ipTablesCollector)
-		collectors = append(collectors, nodeLogsCollector)
-		collectors = append(collectors, kubeletCmdCollector)
-		collectors = append(collectors, systemPerfCollector)
+	enabledCollectorNames := strings.Fields(enabledCollectorString)
+
+	for _, collectorName := range enabledCollectorNames {
+		collectors = append(collectors, allCollectorsByName[collectorName])
 	}
 
-	if contains(collectorList, "OSM") {
-		collectors = append(collectors, osmCollector)
+	return collectors
+}
+
+// selectDiagnosers select the diagnosers to run
+func selectDiagnosers(allDiagnosersByName map[string]interfaces.Diagnoser) []interfaces.Diagnoser {
+	diagnosers := []interfaces.Diagnoser{}
+
+	//read list of diagnosers that are enabled
+	enabledDiagnoserString, found := os.LookupEnv("ENABLED_DIAGNOSERS")
```

Comment:
Are ENABLED_XXX new variables? If yes, is it possible to remove them completely?
We already discussed this point and I think for a first iteration, having
		//if not defined, default to all diagnosers enabled
		enabledDiagnoserString = "networkconfig networkoutbound"
is a what we want.
The user should only have to choose exporters

What do you think? @Tatsinnit

Reply:
Yes new variables - yes we discussed this and I argue that it is a useful capability to support across all collectors / diagnosers / exporters. Arc need collectors to be configurable like this also, and it is only a matter of time until diagnosers need to be as well. No harm in making them default to all on (the user doesn't **have** to choose) but also support overriding like this.

Reply:
I will plus 1 to **not** adding new variables without thinking whole causal effect it will have, especially some of the `osm`, `smi` work is utilising `collector_list` to uniquely identify new work.

* Currently the `collectors/diagnosers` are set of default periscope collectors, as the OSM and SMI (Service Mesh team) added we have added this logic https://github.com/Azure/aks-periscope/blob/master/cmd/aks-periscope/aks-periscope.go#L76 to handle those scenarios.

How adding new variable is needing clarity in this design:

* Breaking change as to what is the default working behaviour?
* Why are we adding these new variable? What is the use case, what requirement needs it as `selectable` option? 

Thanks 🙏

Reply:
Correct, COLLECTOR_LIST is intended to be deprecated. There are no backwards compatibility issues or breaking changes with removing it as far as I am aware, as long as azCLI and VSCode extension haven't yet changed to use it? 

Arc scenario requires being able to only run some collectors, not the full set. Values will be provided via deployment yaml, example here: https://github.com/davidkydd/aks-periscope/blob/0071201dd41cd159e87e5a574bdfbcc271c122b1/deployment/aks-periscope-arc.yaml#L147

Reply:
💡 `Adding this as a thought` : So, adding new configurable `vars`, will become a breaking change for **az-cli and vscode** as well because those `default vars` need to be maintained somehow in most cleanest non scattered manner (not sure how). (some proposal was here) I will think more about this **but** _**I think at this point see this discussion here**_: https://github.com/Azure/aks-periscope/pull/75/files#r661261039 

* Essentially, in current state when we remove the "Collector_List", **and** adding new `vars` it will break things, and any new changes on top will be backward breaking change,  Thanks.

Reply:
My fork retains support for COLLECTOR_LIST for backcompat while adding support for SELECTED_COLLECTORS too. COLLECTOR_LIST is checked first and used preferentially if both are defined. If you want I can bring this code to the PR too, then we can deprecate COLLECTOR_LIST and move VSCode / CLI to use SELECTED_COLLECTORS at our leisure before removing COLLECTOR_LIST support from core periscope?

Reply:
I don't see this issue with backcompat since if we removed this the log would still be collected as before (with some more, but this can't be an issue)

Reply:
💡Also, Going back and possibly asking again why are we adding this again, please? `ENABLED_COLLECTORS, ENABLED_DIAGNOSERS, ENABLED_EXPORTERS.` 

* Is this `ARC` specific requirement which is the key scenario here or is this the refactor?

also cc: @arnaud-tincelin in case you know. Thanks.

Reply:
@arnaud-tincelin the backcompat issue is to do with integration with az-cli and vscode extensions, and would prevent the issue which @Tatsinnit described here being a breaking change: https://github.com/Azure/aks-periscope/pull/83#discussion_r675228034

Yes ENABLED_COLLECTORS is a direct replacement for COLLECTOR_LIST - Arc also have a requirement for a LocalMachineExporter and we need a way to selectively enable and disable this functionality, 

The ability to determine which components (collectors, diagnosers, exporters) are executed via the deployment yaml is a generally useful feature to support, as it allows the user to control the behaviour of periscope at a coarse but clearly defined level. We have already seen from just a few feature requests that it is necessary to be able to control which collectors and exporters are executed, so we should be proactive and enable this capability for diagnosers at the same time. 

Therefore:
We should standardize on naming and either add the three configs:

ENABLED_COLLECTORS
ENABLED_EXPORTERS 
ENABLED_DIAGNOSERS 

or the equivalent:

COLLECTOR_LIST
EXPORTER_LIST
DIAGNOSER_LIST

The form ENABLED_XYZ  is more explicit than XYZ_LIST as to what it actually does, which is why I am suggesting we use the ENABLED_XYZ form - imagine later on if we want to invert this logic for some reason (a very plausible scenario) and add a DISABLED_COLLECTORS config flag, for the requirement that an environment can define what collectors *cannot* run on it, but so that it will automatically include and run any new collectors that are added by default. 

In this scenario COLLECTOR_LIST would become ambiguous as to its function (is it the list of collectors that are enabled, or disabled?)

Reply:
My position is as follow:
- expoters shall be variabilized as they depend on the ecosystem around periscope
- diagnosers and collectors don't have the same kind of dependencies and shall not be variabilized. Periscope provides a bunch of logs, it's up to the user to use them or not, to filter them, sort them etc ... later on
- I see no issue with backcompat, these are env variables. If the code don't use them it won't break anything. Why the need to have these vars anyway?

Reply:
We are going around in circles... Arc scenarios need the ability to disable some of the current collectors. Back compat is for the existing CLI implementation which might make use of COLLECTOR_LIST already (Tats can confirm)